Unhelpful diagnostic on variant containing #[from] #[backtrace] Error, Backtrace

```rust
#![feature(backtrace)]

use std::backtrace::Backtrace;
use thiserror::Error;

#[derive(Error, Debug)]
#[error("...")]
pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
```

```console
error[E0063]: missing field `1` in initializer of `Error`
 --> src/main.rs:8:12
  |
8 | pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
  |            ^^^^^ missing `1`
```

With the following change it compiles successfully, which is expected:

```diff
- pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
+ pub struct Error(#[from] std::io::Error, Backtrace);
```

and with the following change it produces a helpful diagnostic similar to what I think we should expect in the original case:

```diff
- pub struct Error(#[from] std::io::Error, Backtrace);
+ pub struct Error(#[from] std::io::Error, String);
```

```console
error: deriving From requires no fields other than source and backtrace
 --> src/main.rs:8:18
  |
8 | pub struct Error(#[from] std::io::Error, String);
  |                  ^^^^^^^
```

--- a/impl/src/valid.rs
+++ b/impl/src/valid.rs
@@ -180,7 +180,11 @@ fn check_field_attrs(fields: &[Field]) -> Result<()> {
         }
     }
     if let Some(from_field) = from_field {
-        if fields.len() > 1 + has_backtrace as usize {
+        let max_expected_fields = match backtrace_field {
+            Some(backtrace_field) => 1 + !same_member(from_field, backtrace_field) as usize,
+            None => 1 + has_backtrace as usize,
+        };
+        if fields.len() > max_expected_fields {
             return Err(Error::new_spanned(
                 from_field.attrs.from,
                 "deriving From requires no fields other than source and backtrace",

--- a/tests/ui/from-backtrace-backtrace.rs
+++ b/tests/ui/from-backtrace-backtrace.rs
@@ -1,0 +1,12 @@
+// https://github.com/dtolnay/thiserror/issues/163
+
+#![feature(backtrace)]
+
+use std::backtrace::Backtrace;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+#[error("...")]
+pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
+
+fn main() {}

--- a/tests/ui/from-backtrace-backtrace.stderr
+++ b/tests/ui/from-backtrace-backtrace.stderr
@@ -1,0 +1,5 @@
+error[E0063]: missing field `1` in initializer of `Error`
+  --> tests/ui/from-backtrace-backtrace.rs:10:12
+   |
+10 | pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
+   |            ^^^^^ missing `1`

--- a/tests/ui/from-backtrace-backtrace.stderr
+++ b/tests/ui/from-backtrace-backtrace.stderr
@@ -1,5 +1,5 @@
-error[E0063]: missing field `1` in initializer of `Error`
-  --> tests/ui/from-backtrace-backtrace.rs:10:12
+error: deriving From requires no fields other than source and backtrace
+  --> tests/ui/from-backtrace-backtrace.rs:10:18
    |
 10 | pub struct Error(#[from] #[backtrace] std::io::Error, Backtrace);
-   |            ^^^^^ missing `1`
+   |                  ^^^^^^^

